WordPress/gutenberg
# Add Custom Fields and Advanced Panels to Options modal

## What this does

![advanced-panels](https://user-images.githubusercontent.com/612155/47126296-92586880-d2d3-11e8-8a12-3fa407b2ff79.gif)

Closes #10210 and closes #3228.

Firstly, this PR allows the user to enable or disable _Advanced Panels_ (AKA Meta Boxes) via the Screen Options modal that was added in https://github.com/WordPress/gutenberg/pull/10215. 

Secondly, it adds _Custom Fields_ to Gutenberg in the form of an _Advanced Panel_ that is disabled by default.

## How it works

- The title and ID of each Meta Box is stored in the `core/edit-post` Redux store.
- The `MetaBoxTitles` HOC provides an easy way to list all of the registered Meta Boxes within Gutenberg.
- The `MetaBoxesVisibility` and `MetaBoxVisibility` components respond to the existing `isEditorPanelEnabled` selector and show/hide Meta Boxes using `el.style.display = 'none';`
- The Custom Fields Meta Box (`postcustom`) is no longer filtered out by `gutenberg_filter_meta_boxes`

## How to test

1. Create a new post
2. Click on the _More_ menu and select _Options_
3. Enable _Custom Fields_—the _Custom Fields_ meta box should appear and be usable
4. Activate a plugin that adds a Meta Box, e.g. ACF
5. Create a new post
6. You should be able to toggle on/off the panels that were added by the plugin

## Remaining tasks

- [x] Fix issue which causes _Custom Fields_ panel to be enabled after the user upgrades Gutenberg
- [x] Prevent empty `.edit-post-layout__metaboxes` container from appearing when all _Advanced Panels_ are disabled
- [x] Tests and documentation
